beluga/include/beluga/algorithm/raycasting/bresenham.hpp

```diff
@@ -77,22 +77,22 @@ class Bresenham2i {
       using reference = Vector2&;
 
       /// Default constructor.
-      iterator() noexcept = default;
+      iterator() = default;
 
       /// Default copy constructor.
-      iterator(const iterator&) noexcept = default;
+      iterator(const iterator& other) = default;
 
       /// Default move constructor.
-      iterator(iterator&&) noexcept = default;
+      iterator(iterator&&) = default;  // NOLINT(performance-noexcept-move-constructor)
```

Comment:
Removing `noexcept` as it causes implicit deletion of default generated constructors and operator overloads (either older Eigen3 not declaring noexcept constructors for its types or compiler/standard [shenaningans](https://stackoverflow.com/questions/75008386/constructor-is-implicitly-deleted-because-its-exception-specification-does-not-m), not sure).

Reply:
I know it looks crappy on code, but if we want to keep the noexcept whenever possible we can define a macro that replaces into "noexcept" whenever we can use it and nothing elsewhere.

Reply:
> (either older Eigen3 not declaring noexcept constructors for its types or compiler/standard [shenaningans](https://stackoverflow.com/questions/75008386/constructor-is-implicitly-deleted-because-its-exception-specification-does-not-m), not sure)

I followed the rabbit hole just for fun (?). Here is an MRE:
```cpp
#include <type_traits>

template <class Scalar>
class vector {
 public:
  vector() = default;
  vector(vector<Scalar>&&) noexcept(std::is_nothrow_move_constructible<Scalar>::value) {};
};

class iterator {
 public:
  iterator() noexcept = default;
  iterator(iterator&&) noexcept(true) = default;

 private:
  vector<int> vector_;
};

int main() {
  auto it = iterator();
}
```

See https://godbolt.org/z/dfWvbrM4s.

Notes:
- Eigen added `noexcept` move contructors in [`Eigen 3.3-beta2`](http://eigen.tuxfamily.org/index.php?title=ChangeLog#Eigen_3.3-beta2) so this is not it. Commit hash: [`8d97ba6`](https://gitlab.com/libeigen/eigen/-/commit/8d97ba6b2251aabf325ff74f24959ceaa85cf11e).
- The error points towards the compiler not instantiating `std::is_nothrow_move_constructible` before evaluating `iterator`'s move constructor:
  ```bash
  <source>:13:3: note: 'iterator::iterator(iterator&&) noexcept' is implicitly deleted because its exception-specification does not match the implicit exception-specification 'noexcept (<uninstantiated>)'
  ```
- The above example works with [`gcc 5.1`](https://gcc.gnu.org/gcc-5/changes.html) compiling with `-std=c++17`. See https://godbolt.org/z/e74dvKeWr.

Using non-default special member functions is the only workaround I could find that can use `noexcept(true)` with `gcc 4.9.3`: https://godbolt.org/z/Kobbx6cvj

Reply:
could we use the [rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three) instead?

It seems we're defaulting copy/move constructor/assignment, and we don't have a custom destructor.
So we could just not do anything and the compiler will figure out if noexcept can apply or not.

Reply:
> could we use the [rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three) instead?

+1

Reply:
Good idea. Done in https://github.com/Ekumen-OS/beluga/pull/209/commits/19efb3221b06f5fddc7ea3fcbc1cc3ef60f97bc5.